File: game-development-frontend/lib/api/capture.ts
// game-development-frontend/lib/api/capture.ts

import { apiClient } from './client'
import type { GeoLocation, PokemonInstance } from '@/types'

export interface RareCaptureResult {
  success: boolean
  rare_pokemon_id: number
  pokemon_instance: PokemonInstance | null
  remaining_pokeballs: number
}

export interface CommonCaptureResult {
  success: boolean
  pokemon_instance: PokemonInstance | null
  remaining_pokeballs: number
}

export const captureApi = {
  captureRare: (payload: {
    rare_pokemon_id: number
    pokeball_item_id: number
    player_location: GeoLocation
  }) => apiClient.post<RareCaptureResult>('/capture/rare', payload),

  captureCommon: (payload: {
    species_id: number
    level: number
    pokemon_current_hp: number
    pokemon_max_hp: number
    pokeball_item_id: number
    player_location: GeoLocation
  }) => apiClient.post<CommonCaptureResult>('/capture/common', payload),
}

File: app/api/schemas/capture.py
# app/api/schemas/capture.py

from __future__ import annotations

from pydantic import BaseModel

from app.api.schemas.common import GeoLocationModel
from app.api.schemas.pokemon import PokemonInstanceModel


class CaptureRequest(BaseModel):
    rare_pokemon_id: int
    pokeball_item_id: int
    player_location: GeoLocationModel


class CaptureResponse(BaseModel):
    success: bool
    rare_pokemon_id: int
    pokemon_instance: PokemonInstanceModel | None
    remaining_pokeballs: int


class CommonCaptureRequest(BaseModel):
    species_id: int
    level: int
    pokemon_current_hp: int
    pokemon_max_hp: int
    pokeball_item_id: int
    player_location: GeoLocationModel


class CommonCaptureResponse(BaseModel):
    success: bool
    pokemon_instance: PokemonInstanceModel | None
    remaining_pokeballs: int
